src/System.Runtime/tests/System/StringGetHashCodeTests.cs

```diff
@@ -0,0 +1,62 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Xunit;
+
+namespace System.Tests
+{
+    public class StringGetHashCodeTests : RemoteExecutorTestBase
+    {
+        /// <summary>
+        /// Ensure that string hash codes are randomized by getting the hash code for the same string in two processes
+        /// and confirming it is different (modulo possible values of int).
+        /// If the legacy hash codes are being returned, it will not be different.
+        /// </summary>
+        [Fact]
+        public void GetHashCode_UseSameStringInTwoProcesses_ReturnsDifferentHashCodes()
+        {
+            const string abc = "same string in different processes";
+            int childHashCode, parentHashCode = abc.GetHashCode(), timesTried = 0;
+
+            Func<int> GetChildHashCode = () =>
+            {
+                using (RemoteInvokeHandle handle = RemoteInvoke(() => abc.GetHashCode(), new RemoteInvokeOptions { CheckExitCode = false }))
+                {
+                    handle.Process.WaitForExit();
+                    return handle.Process.ExitCode;
+                }
+            };
+
+            do
+            {
+                // very small chance the child and parent hashcode are the same. To further reduce chance of collision we try up to 3 times
+                childHashCode = GetChildHashCode();
+                timesTried++;
+            } while (parentHashCode == childHashCode && timesTried < 3);
+            Assert.NotEqual(parentHashCode, childHashCode);
+        }
+
+        /// <summary>
+        /// Ensure that dictionary changes its comparer to randomized after it encounters more than 100 collisions on a single hash bucket
+        /// </summary>
+        [Fact]
+        public void AddToDictionaryWithNonRandomizedStringComparer_PassHashCollisionThreshold_SwitchesComparerToRandomized()
+        {
+            var dict = new Dictionary<string, int>();
+            Assert.NotSame(EqualityComparer<string>.Default, dict.Comparer);
```

Comment:
This looks like a bug. Dictionary is not supposed to leak the fact that it is internally using the NonRandomizedComparer.

Reply:
The 100 should not be hardcoded (as @danmosemsft suggested in #24466).
The test belongs to Dictionary tests, not here IMO.
You should ensure that the hardcoded values really cause collisions. (what if someone changes the algorithm?)

The bigger question: Do we want this test at all? -- @danmosemsft's intention in #24466 was to have a test to make sure we use randomized hash code (at certain point) - which sounds reasonable.
However, I don't like the fact we have to hardcode the hash code algorithm in the test itself (via precomputed collisions). That is brittle and may not work across runtimes (you may need another table for .NET Framework as you found out from NetFx run failure per offline discussion). At minimum, the test should check, the values are still causing collisions and fail otherwise ...

@jkotas are you suggesting to always use RandomizedComparer in Dictionary? Not just after 100 collisions?

Reply:
> @jkotas are you suggesting to always use RandomizedComparer in Dictionary? Not just after 100 collisions?

> @jkotas are you suggesting to always use RandomizedComparer in Dictionary? Not just after 100 collisions?

I am suggesting that Dictionary should not be leaking the details of its internal perf optimization through its public surface. I think that the following test should pass:

```
var dict = new Dictionary<string,int>(EqualityComparer<string>.Default);
Console.WriteLine(dict.Comparer.Equals(EqualityComparer<string>.Default) ? "PASS" : "FAIL");
```

It fails on .NET Core master. It would pass on .NET Core 2.0 - we have regressed the behavior since .NET Core 2.0.

Reply:
@karelz I moved the collision-checking test to Dictionary test class

@jkotas I modified the collision-checking test to reflect that in full-framework the dictionary comparer is randomized by default (as you mentioned above), and otherwise testing after reaching collision threshold it switches to randomized.

@jkotas Regarding your comment on how the internal comparer type is exposed, do you want me to add a new issue/bug explaining this so that gets tracked separately?

Reply:
> do you want me to add a new issue/bug explaining

Yes please. Also, your test should not be taking advantage of this bug.